### DISCUSSION TITLE
is it OK to make a default feature which is not available on platforms / configurations?

### DISCUSSION THREAD

**Discussion by ilya-lavrenov** [microsoft/vcpkg#31684] [General]
We have a product with plugin architecture where plugin are responsible for devices on your system. 
E.g. on macOS our GPU plugin does not support Apple's GPU, so we are not building it here. While on Linux / Windows the plugin is available and we usually build it by default.

With vcpkg, I have to add `supports` condition for `gpu` feature to make it unavailable for macOS. Also, I have added this feature to be default, which by our logic mean "build GPU plugin by default on platforms where it's available".  But, when I try to build package for macOS, it reports that one of the default features are not available. 

What is the proper way to handle this?
I hope that this is normal "include some features by default on configurations where it's available". E.g. by default cmake can discover some libraries installed on the system and build additional functionality in cross-platform way, while with vcpkg such behavior does not look cross-platform, because users have to explicitly disable GPU feature on macOS.

The similar questions with dependencies. We are using xbyak which is not available for ARM platforms and it's explicitly stated in xbyak vcpkg.json file. So, when we include xbyak as a dependency, on macOS with Apple Silicon vcpkg reports that dependency is not available and I have to copy `supports` content of xbyak to ours `platforms`. How to normally handle such situations? For large projects it seems quite stupid to copy-paste limitations of dependencies. Is it worth to introduce some option to cover "I want to install dependency for platform where it's available"? It looks logical in most of cases, while for cases when I really want to limit installation condition for dependency, I can use current behavior.

Last question: I see the `onednn` library is ported to vcpkg, but really it has a lot of alternative options (backends, not features in terms of vcpkg). Default options do not work for us, because we use different backends, so is it OK to download `onednn` explicitly in our port (assuming we will publish our port to this repo later) and build with options suitable for our project? Note, that we will do it only for dynamic build, so conflicts are impossible when `onednn` is used multiple times in a dependency tree of some high-level application.

**Comment by dg0yt** [comment]
For the first question:
At the moment, default features need to work on all platforms. (We are waiting for https://github.com/microsoft/vcpkg-tool/pull/813.).
The current workaround is to have a feature `default-features`. It can pull in other features via dependencies wth `platform` condition.
